maint/local_python/binding_c.py

```diff
@@ -584,6 +675,50 @@ def dump_copy_right():
     G.out.append(" */")
     G.out.append("")
 
+def dump_profiling_qmpi(func, mapping):
+    parameters = ""
+    for p in func['c_parameters']:
+        parameters = parameters + ", " + p['name']
+
+    func_name = get_function_name(func, mapping)
+    s = get_declare_function(func, mapping)
+
+    G.out.append("#ifdef ENABLE_QMPI")
+    dump_line_with_break(s)
+    G.out.append("{")
+    G.out.append("    QMPI_Context context;")
+    G.out.append("    Q%s_t *fn_ptr;" % (func_name))
+    G.out.append("")
+    G.out.append("    context.storage_stack = NULL;")
+    G.out.append("")
+    if func_name == "MPI_Pcontrol":
+        G.out.append("    va_list varargs;")
+        G.out.append("    va_start(varargs, level);")
+        G.out.append("")
+    if "initcheck" in func['skip']:
+        G.out.append("    /* Initialize the tools that have been registered */")
+        G.out.append("    for (int i = 1; i < MPIR_QMPI_num_tools + 1; i++) {")
+        G.out.append("        MPIR_QMPI_tool_init_callbacks[i](i);")
+        G.out.append("    }")
```

Comment:
If `MPII_qmpi_setup` never been called yet, wouldn't ` MPIR_QMPI_tool_init_callbacks` unavailable? In fact, I don't understand why do you need this block at all. Isn't calling `MPII_qmpi_setup()` sufficient? I mean, it will call the init_callbacks inside if it's first time, right?

Reply:
If `MPII_qmpi_setup` hasn't been called, that means no tools were registered and `MPIR_QMPI_num_tools` will be `0`. So this loop won't execute.

Reply:
Ahh, I see. That is non-obvious. At least need place the `for-`loop inside the `if-` clause to make the above understanding explicit.

But still --
> In fact, I don't understand why do you need this block at all. Isn't calling MPII_qmpi_setup() sufficient? I mean, it will call the init_callbacks inside if it's first time, right?

Reply:
No, the callback functions are the ones registered by the tool to be called when all of the tools have finished registering. This allows the tools to know the order of all of the tools and cache the "next" functions for all of the functions that they intercept.

This whole thing can benefit from some more formal text describing all of the functions, but my plan to was to write that up for the Tools Working Group when I finished getting this merged. Maybe I can put together a slide deck or something in the meantime. 😄